[WIP] Creating Docker images with pure Salt and running Salt commands inside containers

dmacvicar opened the pull request:
### What does this PR do?
- It implements creating Docker images from sls states (Salt on the image not required, base image only needs python)

![screenshot from 2016-07-05 23-16-16](https://cloud.githubusercontent.com/assets/15332/16617553/9c8dad00-4384-11e6-8641-cf520c345b9f.png)

![cmds_bvwyaairc-](https://cloud.githubusercontent.com/assets/15332/16617533/7bfb53bc-4384-11e6-8110-87f1e9792546.jpg)
- Allows to run arbitrary Salt modules inside a container without needing Salt in the container (uses salt-thin).

![cmrl4paviaemgly](https://cloud.githubusercontent.com/assets/15332/16617549/8e29c4ba-4384-11e6-98cb-ff6a33e24548.jpg)
- It was a [Hackweek 14 Project](https://hackweek.suse.com/14/projects/1756) at [SUSE](http://www.suse.com).
- Lightning talk: [![default](https://cloud.githubusercontent.com/assets/15332/16732477/b6ea7aaa-477d-11e6-9fd6-3b697e7a87a5.jpg)](https://www.youtube.com/watch?v=2znjgf9Q7J0)
### What issues does this PR fix or reference?
- https://www.redhat.com/en/about/press-releases/red-hat-launches-ansible-native-container-workflow-project
### Tests written?

Yes. (dockerng.sls pending)

cc @meaksh @isbm @dincamihai

dmacvicar committed:
first version of building docker images

dmacvicar committed:
test return value

dmacvicar committed:
testcase for call

dmacvicar committed:
escape salt-call command

dmacvicar committed:
make sure function parameter is there

dmacvicar committed:
move import to the top

dmacvicar committed:
implement keyword arguments for dockerng.call

cachedout commented:
This is really awesome!

thatch45 commented:
Awesome! You got it in @dmacvicar !

rallytime commented:
Neat!

@dmacvicar There are a couple of [pylint errors](https://jenkins.saltstack.com/job/PR/job/salt-pr-lint-n/3100/violations/file/salt/modules/dockerng.py/) to clean up. It would also be good to get some `.. versionadded:: Carbon` tags in the docstrings of each of the new public-facing functions.

isbm commented on salt/modules/dockerng.py:
```diff
@@ -5529,3 +5541,214 @@ def script_retcode(name,
+    return st_.state.compile_high_data(high_data)
+
+
+def _gather_pillar(pillarenv, pillar_override, grains={}):
```
Just change the signature like this:
```python
def _gather_pillar(pillarenv, pillar_override, **grains):
```

dmacvicar committed:
pylint: W0611(unused-import)

dmacvicar committed:
pylint: W8391(blank-line-at-end-of-file)

dmacvicar committed:
pylint: W0102(dangerous-default-value)

dmacvicar committed:
add versionadded:: Carbon to new public functions

dmacvicar commented:
Thanks for the feedback. I corrected everything mentioned until now.

* What do you think of the name of the `call` function? Is there any convention to these meta-calls? snapper module will introduce a similar one. It would be cool if they follow some convention in naming and style.

Originally, the function signature was: `call(container_id, function=None, args=[], kwargs={})` but now is just `call(container_id, function, *args, **kwargs), which means I just forward all parameters. It makes it easier to use but also means that I have to filter `__key` like args injected by salt so that `call(id, 'test.ping') works.

* The current thin handling is a bit brute force and does not has the amount of logic salt-ssh does to the unpacking.

dmacvicar commented:
- [ ] Delete the tarballs in /tmp before commit

isbm commented:
@dmacvicar The `call` signature is now correct. Nothing wrong with the filtering, since everything starting with `__` is usually technical "carryover" that you can throw away, in case you need.

isbm commented:
@dmacvicar The name `call` itself isn't really bad, although keep in mind that there is `__call__` function to every "callable" class:
```python
class Foo(object):
    def __call__(self):
        return "Hello world"
print(Foo()())
```

isbm commented on salt/modules/dockerng.py:
```diff
@@ -5529,3 +5540,216 @@ def script_retcode(name,
+    '''
+    if 'cp.fileclient' not in __context__:
+        __context__['cp.fileclient'] = \
+                salt.fileclient.get_file_client(__opts__)
```
BTW, this can be one line. No need to _enforce_ 80 characters in Salt.

thatch45 replied:
yes, this can be one line

dmacvicar replied:
Fixed.
As I showed @isbm, main problem is that running ` pylint --rcfile=.testing.pylintrc --disable=W1307 salt/modules/dockerng` does not have the same behavior as what I see in Jenkins.

isbm commented on salt/modules/dockerng.py:
```diff
@@ -5529,3 +5540,216 @@ def script_retcode(name,
+    high_data, errors = st_.render_highstate({saltenv: mods})
+    high_data, ext_errors = st_.state.reconcile_extend(high_data)
+    errors += ext_errors
+    errors += st_.state.verify_high(high_data)
```
Up to you though, but usually I try always avoid `+=` on strings, replacing by:
```
errors = [errors]
errors.append(ext_errors)
errors.append(st_.state.verify_high(high_data))
if errors:
    return ' '.join(errors)
```

And so on. But exactly this part isn't very much performant, so you can leave it as is.

isbm commented on salt/modules/dockerng.py:
```diff
@@ -5529,3 +5540,216 @@ def script_retcode(name,
+    errors += ext_errors
+    errors += st_.state.verify_high(high_data)
+    if errors:
+        return errors
```
Up to you though, but usually I try always avoid `+=` on strings, replacing by:
```
errors = [errors, ext_errors, st_.state.verify_high(high_data)]
if errors:
    return ' '.join(errors)
```

And so on. But exactly this part isn't very much performant, so you can leave it as is.

thedrow replied:
I think that joining will be slower than +=. Also whenever you do something like this, use tuples.

isbm replied:
@thedrow Actually, I just wrote a little script to test both, getting numbers. Joining is about 3x times faster. So it didn't changed since Python 1.5 version :wink:

isbm commented on salt/modules/dockerng.py:
```diff
@@ -5529,3 +5540,216 @@ def script_retcode(name,
+    '''
+    # put_archive reqires the path to exist
+    ret = __salt__['dockerng.run_all'](name, 'mkdir -p /tmp/salt_thin')
+    if ret['retcode'] != 0:
```
Nitpick: in this case `!= 0` is only an explicit self-documentation "Not Zero" for a code reader. But technically you can say just:
```python
if ret['retcode']:
    ....
```
To a Python reader it means "something non-False is in the "retcode" section".

isbm commented on salt/modules/dockerng.py:
```diff
@@ -5529,3 +5540,216 @@ def script_retcode(name,
+        '-l', 'quiet',
+        '--',
+        function
+    ] + list(args) + ['{0}={1}'.format(key, value) for (key, value) in kwargs.items() if not key.startswith('__')]
```
Few things:

1. Here `args` already a list.
2. Use `six.iteritems` instead of `dict.items()`: https://pythonhosted.org/six/#six.iteritems

dmacvicar replied:
`args` is a tuple, I converted it to list because I got an error from the interpreter.

isbm replied:
Ah, right. Immutable.

s0undt3ch replied:
And use `six.iteritems()` since `obj.iteritems()` is not python 3 compatible.
Also, `.items()` is not wrong, it just uses more memory under python 2(On python 3 its equivalent to `obj.iteritems()`.

isbm commented on salt/modules/dockerng.py:
```diff
@@ -5529,3 +5540,216 @@ def script_retcode(name,
+
+    .. code-block:: bash
+
+        salt myminion dockerng.sls compassionate_mirzakhani mods=rails,web
```
:wink:

isbm commented on salt/modules/dockerng.py:
```diff
@@ -5529,3 +5540,216 @@ def script_retcode(name,
+    if mods is not None:
+        mods = [x.strip() for x in mods.split(',')]
+    else:
+        mods = []
```
This can be one line:
```python
mods = [item.strip() for item in mods.split(',') if mods else []]
```

isbm commented on salt/modules/dockerng.py:
```diff
@@ -5529,3 +5540,216 @@ def script_retcode(name,
+    try:
+        trans_tar_sha256 = salt.utils.get_hash(trans_tar, 'sha256')
+        __salt__['dockerng.copy_to'](name, trans_tar,
+                                     '/tmp/salt_state.tgz',
```
Here I would suggest to use a random prefix in the archive name.

thatch45 replied:
subdir again, this is almost as bad as downloading code from the internet without validation and then executing it ;)

dmacvicar commented on salt/modules/dockerng.py:
```diff
@@ -5529,3 +5540,216 @@ def script_retcode(name,
+    grains = __salt__['dockerng.call'](name, function='grains.items')
+
+    # compile pillar with container grains
+    pillar = _gather_pillar(saltenv, {}, grains)
```
I think this call is wrong now that I changed the signature. I think I have to pass `**grains`here to convert the dict to keyword arguments.

isbm replied:
Well, yes. If you have something like:
```python
def foo(*args, **kwargs): pass
```
Then you supposed to call it this way:
```python
foo(**['some', 'stuff'], **{'here': 'after'})
```

isbm commented on salt/modules/dockerng.py:
```diff
@@ -5529,3 +5540,216 @@ def script_retcode(name,
+    ret = __salt__['dockerng.create'](image=base,
+                                      cmd='/usr/bin/sleep infinity',
+                                      interactive=True, tty=True)
+    id = ret['Id']
```
JFYI: `id` is a built-in function and you are over-writing it.

thatch45 replied:
our lint checks should get mad about this one, we often just use the var `id_`

thatch45 commented on salt/modules/dockerng.py:
```diff
@@ -5529,3 +5540,216 @@ def script_retcode(name,
+
+    '''
+    # put_archive reqires the path to exist
+    ret = __salt__['dockerng.run_all'](name, 'mkdir -p /tmp/salt_thin')
```
this needs to go in a random prefix subdir, otherwise this gets us a CVE

dmacvicar replied:
I thought about that but I could not think of the attack. The user executing the stuff into the container is the same root user of the host. Yes, you could escalate if you already broke into the root account. Can you explain me the attack?

In any case, we can make it a random subdir, but we have to think how "caching" work in this regard. `dockerng.copy_to` keeps the thin there if it was already there, so subsequent commands will not copy it again.

Would also storing the random path in the module state (is there any state? like __context__?) work?

thatch45 commented on salt/modules/dockerng.py:
```diff
@@ -5529,3 +5540,216 @@ def script_retcode(name,
+    # move salt into the container
+    thin_path = salt.utils.thin.gen_thin(__opts__['cachedir'])
+    with io.open(thin_path, 'rb') as file:
+        _client_wrapper('put_archive', name, '/tmp/salt_thin', file)
```
Again, needs to be in a random named subdir, and that subdir can't be world writable

thatch45 commented on salt/modules/dockerng.py:
```diff
@@ -5529,3 +5540,216 @@ def script_retcode(name,
+
+    salt_argv = [
+        'python',
+        '/tmp/salt_thin/salt-call',
```
subdir

thatch45 commented on salt/modules/dockerng.py:
```diff
@@ -5529,3 +5540,216 @@ def script_retcode(name,
+
+        # Now execute the state into the container
+        ret = __salt__['dockerng.call'](name, function='state.pkg',
+                                        args=['/tmp/salt_state.tgz',
```
random subdir

thatch45 commented:
so the thing with the random subdirs is because an unprivileged process in the container could modify the code that exists in /tmp before it is executed by salt, then their code would run with salt's privs, therefore allowing someone to steal the container from an unprivileged process.

dmacvicar committed:
pylint: E1121(too-many-function-args)

dmacvicar commented:
Ok. I think the problem is not that the unprivileged user modifies the code but that the code exists before, as `copy_to` by default not overwrite it.

In any case I think there is no point in caching as we have to prevent the thin to be committed into any intermediate image, so I will have to copy the thin on *every command*. This will be inefficient but I don't have an alternative for now.

s0undt3ch commented:
Mount and unmount a partition which only has the the thin?

dmacvicar commented:
@s0undt3ch my original prototype used `docker build`  API with a generated Dockerfile, and `docker build` this not support mounting.

Then I switched to starting a container myself, and the initial version did a mount. For building images I could mount the partition at the start, but when I refactored the ability to run states and modules directly in random containers (nice idea from @thatch45 ), Docker does not allow to add volume mounts to running containers, then I changed the approach to copying.

thedrow commented:
You can stop the container and start it with a mounted volume.
@thatch45 Lemme review all this during the weekend please :)

thedrow commented:
How do you run a base.sls file with this and does this commit the container whenever a state is successfully applied?

thatch45 commented:
@dmacvicar 
Here is the CWE:
https://cwe.mitre.org/data/definitions/377.html

Basically stated, whether an attack can be illustrated or not the use of predictable directories in /tmp due to the perms on /tmp (1777) would allow said predictable directory to be created before by a malicious attacker and files inserted, even if we delete the directory before code deployment there is still logically enough time for the attack to take place.

So basically said, weather we can illustrate a security vulnerability or not, this would be presented as a vulnerability, so we might as well avoid it.

We do have code in salt-ssh that creates predictable (by us and not an attacker by using data from the target and the originator) to determine the directory, but given the fact that this is all local IO I am far less worried about implementing perfect cache coherency like salt-ssh has.

So all in all, I think that the right thing to do is copy into a random temp dir for now, since that does allow us to modify running containers and can be done securely. Basically, it is an easy thing to do that saves us the risk of any issues :)

thatch45 commented:
@thedrow I don't think this will be merged before the weekend is over, so you should be fine. My main argument here with restarting the container is that we would like to run container scans etc., while the container is running. This would allow us to do things like execute hubblestack on running salt containers and present security reports on the environments of said containers (as one simple example). So I am hesitant to change this from the approach of copying in.

dmacvicar commented:
I will fix all these issues over the weekend or next week. Also I want to
investigate if there are more options.

On Thu, 7 Jul 2016 17:45 Thomas S Hatch, <notifications@github.com> wrote:

> @thedrow <https://github.com/thedrow> I don't think this will be merged
> before the weekend is over, so you should be fine. My main argument here
> with restarting the container is that we would like to run container scans
> etc., while the container is running. This would allow us to do things like
> execute hubblestack on running salt containers and present security reports
> on the environments of said containers (as one simple example). So I am
> hesitant to change this from the approach of copying in.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/saltstack/salt/pull/34484#issuecomment-231119569>, or mute
> the thread
> <https://github.com/notifications/unsubscribe/AAA75Pk6ORdDS0vgvTWbasom5EOsmRQpks5qTR8SgaJpZM4JGBYQ>
> .
>

cachedout commented:
I had to do the following to get this to work:

```
diff --git a/salt/modules/dockerng.py b/salt/modules/dockerng.py
index 76bcd9c..23274a4 100644
--- a/salt/modules/dockerng.py
+++ b/salt/modules/dockerng.py
@@ -5563,7 +5563,7 @@ def _prepare_trans_tar(mods=None, saltenv='base', pillar=None):
     _mk_fileclient()
     trans_tar = salt.client.ssh.state.prep_trans_tar(
         __context__['cp.fileclient'],
-        chunks, refs, pillar=pillar)
+        chunks, refs, pillar=pillar, id_=__opts__['id'])
     return trans_tar
 
 
@@ -5706,9 +5706,9 @@ def sls(name, mods=None, saltenv='base', **kwargs):
                                      overwrite=True)
 
         # Now execute the state into the container
-        ret = __salt__['dockerng.call'](name, function='state.pkg',
-                                        args=['/tmp/salt_state.tgz',
-                                              trans_tar_sha256, 'sha256'])
+        ret = __salt__['dockerng.call'](name, 'state.pkg',
+                                        '/tmp/salt_state.tgz',
+                                              trans_tar_sha256, 'sha256')
 
         # set right exit code if any state failed
         __context__['retcode'] = 0
```

Once it did though, it's totally awesome!

thedrow commented on salt/modules/dockerng.py:
```diff
@@ -5529,3 +5540,216 @@ def script_retcode(name,
+    # be somewhere else
+    refs = salt.client.ssh.state.lowstate_file_refs(chunks)
+    _mk_fileclient()
+    trans_tar = salt.client.ssh.state.prep_trans_tar(
```
Why do we use ssh instead of docker's CP? I don't think requiring an SSH agent on the container is reasonable.

dmacvicar replied:
We are not using ssh. We are only reusing some salt-ssh utility functions to prepare the pre-compiled state. We do use docker cp functions to put that pre-compilled state/pillar bundle and salt-thin. Those salt-ssh functions could be factored out, but that is out of the scope of this PR.

dmacvicar replied:
@cachedout Sorry I refactored the signature of dockerng.call and did not commit everything. Fixed now.

cachedout replied:
@dmacvicar Figured as much. I'm going to keep testing this today. It's really cool!

thedrow commented on salt/modules/dockerng.py:
```diff
@@ -5529,3 +5540,216 @@ def script_retcode(name,
+
+        # set right exit code if any state failed
+        __context__['retcode'] = 0
+        for _, state in ret.iteritems():
```
Why not just six.itervalues?

thedrow commented on salt/modules/dockerng.py:
```diff
@@ -5529,3 +5540,216 @@ def script_retcode(name,
+        __context__['retcode'] = 0
+        for _, state in ret.iteritems():
+            if not state['result']:
+                __context__['retcode'] = 1
```
I think you can write this in a one liner:
```python
__context__['retcode'] = 1 if any(not state['result'] for state in six.itervalues(ret)) else 0
```

thedrow commented on salt/modules/dockerng.py:
```diff
@@ -5529,3 +5540,216 @@ def script_retcode(name,
+            if not state['result']:
+                __context__['retcode'] = 1
+    finally:
+        os.remove(trans_tar)
```
What would happen here if you get a filesystem error for some reason? Let's log it and fail the run.

dmacvicar replied:
Would't that just end in a Exception that would blow the whole call and be translated into a CommandExecutionError?

isbm replied:
You will get an OSError. Then you will get a traceback in logs all around. CLI won't look nice for this.

dmacvicar committed:
use a random path for the thin and clean it up after the call

dmacvicar committed:
adapt to new dockerng.call signature

dmacvicar committed:
use a random path for the trans and clean it up after the call

dmacvicar committed:
fix function name in docs

dmacvicar committed:
rename id to id_

dmacvicar committed:
fix trans tar path

dmacvicar commented:
@thedrow What do you mean with base.sls? There are 3 functions:

* `dockerng.call` : just executes a salt module into a running container
* `dockerng.sls` : from a list of sls modules, calculates the highstate, and executes it into a running container (taking care of pillar data and grains). It does not commit, you can do that with dockerng.commit.
* `dockerng.sls_build` : from a list of sls modules, a base image and a new image name, creates a new image that results of applying the state to the base image. This commits the new image and tags it with the given name.

dmacvicar committed:
turn loop into one liner

dmacvicar committed:
handle os.remove exception

dmacvicar committed:
turn if into a one-liner

dmacvicar commented:
I think all feedback until now is incorporated. Feel free to point to more issues. Some stuff that can be reviewed:

* handling of environments, pillar overrides.

cachedout commented on salt/modules/dockerng.py:
```diff
@@ -5529,3 +5541,244 @@ def script_retcode(name,
+    _mk_fileclient()
+    trans_tar = salt.client.ssh.state.prep_trans_tar(
+        __context__['cp.fileclient'],
+        chunks, refs, pillar=pillar)
```
We still need an id passed in here. Otherwise we're getting this:

```
AttributeError: 'NoneType' object has no attribute 'startswith'
Traceback (most recent call last):
  File "/usr/bin/salt-call", line 11, in <module>
    salt_call()
  File "/home/mp/devel/salt/salt/scripts.py", line 352, in salt_call
    client.run()
  File "/home/mp/devel/salt/salt/cli/call.py", line 58, in run
    caller.run()
  File "/home/mp/devel/salt/salt/cli/caller.py", line 134, in run
    ret = self.call()
  File "/home/mp/devel/salt/salt/cli/caller.py", line 197, in call
    ret['return'] = func(*args, **kwargs)
  File "/home/mp/devel/salt/salt/modules/dockerng.py", line 5780, in sls_build
    __salt__['dockerng.sls'](id_, mods, saltenv, **kwargs)
  File "/home/mp/devel/salt/salt/modules/dockerng.py", line 5711, in sls
    trans_tar = _prepare_trans_tar(mods=mods, saltenv=saltenv, pillar=pillar)
  File "/home/mp/devel/salt/salt/modules/dockerng.py", line 5573, in _prepare_trans_tar
    chunks, refs, pillar=pillar)
  File "/home/mp/devel/salt/salt/client/ssh/state.py", line 148, in prep_trans_tar
    cachedir = os.path.join('salt-ssh', id_)
  File "/usr/lib/python2.7/posixpath.py", line 68, in join
    if b.startswith('/'):
AttributeError: 'NoneType' object has no attribute 'startswith'
Traceback (most recent call last):
  File "/usr/bin/salt-call", line 11, in <module>
    salt_call()
  File "/home/mp/devel/salt/salt/scripts.py", line 352, in salt_call
    client.run()
  File "/home/mp/devel/salt/salt/cli/call.py", line 58, in run
    caller.run()
  File "/home/mp/devel/salt/salt/cli/caller.py", line 134, in run
    ret = self.call()
  File "/home/mp/devel/salt/salt/cli/caller.py", line 197, in call
    ret['return'] = func(*args, **kwargs)
  File "/home/mp/devel/salt/salt/modules/dockerng.py", line 5780, in sls_build
    __salt__['dockerng.sls'](id_, mods, saltenv, **kwargs)
  File "/home/mp/devel/salt/salt/modules/dockerng.py", line 5711, in sls
    trans_tar = _prepare_trans_tar(mods=mods, saltenv=saltenv, pillar=pillar)
  File "/home/mp/devel/salt/salt/modules/dockerng.py", line 5573, in _prepare_trans_tar
    chunks, refs, pillar=pillar)
  File "/home/mp/devel/salt/salt/client/ssh/state.py", line 148, in prep_trans_tar
    cachedir = os.path.join('salt-ssh', id_)
  File "/usr/lib/python2.7/posixpath.py", line 68, in join
    if b.startswith('/'):
AttributeError: 'NoneType' object has no attribute 'startswith'
```

dmacvicar replied:
@cachedout what do you mean by "here" and how do I reproduce this?

cachedout replied:
This was my command: `sudo salt-call --local -ldebug dockerng.sls_build base=centos vim mods=demo `. It seem to blow up if `prep_trans_tar` isn't given an `id` kwarg.

dmacvicar replied:
I do the same. Can you share your demo module?

cachedout replied:
```notify_me:
  test.show_notification:
    - text: 'Notify me'
```

thedrow commented:
I meant an sls file that acts like a highstate but for containers.

dmacvicar commented:
@thedrow a highstate is at the end resolved as a list of sls modules, so I guess this could be added as a layer on top, just like the functions `dockerng.call` `dockerng.sls` and `dockerng.sls_build`are layered one on top of another.

I played with some ideas around it, but none worked very well, but a `dockerng.highstate`should be entirely possible.

dmacvicar committed:
improve error handling to be more consistent

dmacvicar committed:
test that fix /tmp is not used

dmacvicar committed:
typo

dmacvicar committed:
pylint: W1699(incompatible-py3-code)

dmacvicar commented:
@cachedout I am still unable to reproduce the id_ problem.

My setup is Salt master+minion with Vagrant, Salt 2015.08 and the dockerng plugin symlinked into the _modules of the Vagrant shared folder.

How is your environment setup?

I would like to reproduce it to make sure there are no other issues of missing parameters as well.

dmacvicar committed:
do not crash on prep_trans_tar because of missing id_ used to build cachedir

dmacvicar commented:
@cachedout reproduced and fixed. It only happens on 2016.x.

dmacvicar committed:
no need to break line here

dmacvicar commented:
This should be more or less ready for merging.

dmacvicar commented:
As next steps (future work) I suggest:

* creating an equivalent state functions like `dockerng.image_present`

May be reuse the same function and instead of

```yaml
myuser/myimage:mytag:
  dockerng.image_present:
    - build: /home/myuser/docker/myimage
```

also allow

```yaml
myuser/myimage:mytag:
  dockerng.image_present:
    - mods:
      - foo
   - base: leap
```

(I thought about having `sls_build` in the same function but I was not convinced that the amount of possible combinations of parameters was worth it, feedback welcome)

cachedout commented:
Merged! Thanks @dmacvicar. This is really great.

vutny commented on salt/modules/dockerng.py:
```diff
@@ -5529,3 +5541,252 @@ def script_retcode(name,
+
+    .. code-block:: bash
+
+        salt myminion dockerng.call test.ping
```
@dmacvicar It should be like this or maybe I figured it out wrong way:
```
salt myminion dockerng.call containter_name test.ping
```

I think container name or hash argument (`name`) has missed.

vutny commented:
@dmacvicar This is super cool awesome stuff! Thanks!
But it would be really helpful to have some tutorial with examples on docs.saltstack.com about how to use these new `dockerng` functions.

dmacvicar commented:
@vutny I put examples in each call:

https://docs.saltstack.com/en/develop/ref/modules/all/salt.modules.dockerng.html#salt.modules.dockerng.sls_build

anything you are missing?

thedrow commented:
I think we need to add a new section to the to the tutorials page.

thatch45 commented:
I agree, having a tutorial helps usage immensely. I will get that ball rolling

vutny commented:
At list some short intro at the top docstring of `dockerng` module would be great.

@dmacvicar I left an in-line comment above: https://github.com/saltstack/salt/pull/34484#discussion_r72585270

Could you please take a look?
